### ISSUE TITLE
After account update with new access keys, old access keys are valid for next 5min ~

### ISSUE THREAD

**Issue by anandhu-karattu** [noobaa/noobaa-core#8071]
### Environment info
- NooBaa Version: **noobaa-core-5.15.3-20240514.el9.x86_64**
- Platform: **Standalone Noobaa**

### Actual behaviour
After account update with new access keys, old access keys are valid for next 5min ~

### Expected behaviour
After successful account update, the old access keys should become invalid immediately.

### Steps to reproduce
1. Create an account and a bucket. Upload an object in to bucket.
```
[root@3node-x-worker1 ~]# noobaa-cli account add --name account-first --uid 1002 --gid 1002 --fs_backend GPFS --new_buckets_path /mnt/fs1/account-first 2>/dev/null
{
  "response": {
    "code": "AccountCreated",
    "reply": {
      "_id": "664f1b4ce22980ca61b1d181",
      "name": "account-first",
      "email": "account-first",
      "creation_date": "2024-05-23T10:32:44.775Z",
      "access_keys": [
        {
          "access_key": "YIxrhiDGAjJjyqwnxp9I",
          "secret_key": "Md+P1NYohQfK26kKD9guYTKFE1+OJ1DZ3mzQClaF"
        }
      ],
      "nsfs_account_config": {
        "uid": 1002,
        "gid": 1002,
        "new_buckets_path": "/mnt/fs1/account-first",
        "fs_backend": "GPFS"
      },
      "allow_bucket_creation": true,
      "master_key_id": "664e107282dba0d07d51b455"
    }
  }
}
```

```
[root@3node-x-worker1 ~]# mms3 bucket create bucket-first --accountName account-first --filesystemPath /mnt/fs1/account-first/bucket-first
Bucket bucket-first created successfully
```

Step2: Update the account's access key. Only last one character is updated.

```
[root@3node-x-worker1 ~]# noobaa-cli account update --name account-first --fs_backend GPFS --access_key YIxrhiDGAjJjyqwnxp9L --secret_key Md+P1NYohQfK26kKD9guYTKFE1+OJ1DZ3mzQClaF 2>/dev/null
{
  "response": {
    "code": "AccountUpdated",
    "reply": {
      "_id": "664f1b4ce22980ca61b1d181",
      "name": "account-first",
      "email": "account-first",
      "creation_date": "2024-05-23T10:32:44.775Z",
      "access_keys": [
        {
          "access_key": "YIxrhiDGAjJjyqwnxp9L",
          "secret_key": "Md+P1NYohQfK26kKD9guYTKFE1+OJ1DZ3mzQClaF"
        }
      ],
      "nsfs_account_config": {
        "uid": 1002,
        "gid": 1002,
        "new_buckets_path": "/mnt/fs1/account-first",
        "fs_backend": "GPFS"
      },
      "allow_bucket_creation": true,
      "master_key_id": "664e107282dba0d07d51b455"
    }
  }
}
```

**Step3 : Now do the object upload using old access Key : Success. --> This is the issue**
```
[root@anan-rhel921 ~]# AWS_ACCESS_KEY_ID=YIxrhiDGAjJjyqwnxp9I AWS_SECRET_ACCESS_KEY=Md+P1NYohQfK26kKD9guYTKFE1+OJ1DZ3mzQClaF aws --endpoint https://10.11.71.87:6443 --no-verify-ssl s3 cp awscliv2.zip s3://bucket-first
upload: ./awscliv2.zip to s3://bucket-first/awscliv2.zip
```

Step4 : Now verify that new access key is also working 

```
[root@anan-rhel921 ~]# AWS_ACCESS_KEY_ID=YIxrhiDGAjJjyqwnxp9L AWS_SECRET_ACCESS_KEY=Md+P1NYohQfK26kKD9guYTKFE1+OJ1DZ3mzQClaF aws --endpoint https://10.11.71.87:6443 --no-verify-ssl s3 cp awscliv2.zip s3://bucket-first
upload: ./awscliv2.zip to s3://bucket-first/awscliv2.zip
```

Step5 : After few mins, it started erroring out about the invalid key, if I use the old key. This is expected

```
[root@anan-rhel921 ~]# AWS_ACCESS_KEY_ID=YIxrhiDGAjJjyqwnxp9I AWS_SECRET_ACCESS_KEY=Md+P1NYohQfK26kKD9guYTKFE1+OJ1DZ3mzQClaF aws --endpoint https://10.11.71.87:6443 --no-verify-ssl s3 cp awscliv2.zip s3://bucket-first
upload failed: ./awscliv2.zip to s3://bucket-first/awscliv2.zip An error occurred (InvalidAccessKeyId) when calling the CreateMultipartUpload operation: The AWS access key Id you provided does not exist in our records.
```

**Comment by romayalon** [comment]
Adding some internal info - 
When calling `noobaa-cli account update` the account's config file is being updated.
NooBaa service maintains an account cache with a default expiry_ms = 10 minutes, after the expiry the old account won't be valid anymore. 
NooBaa CLI and NooBaa service are different processes, we need to come up with a way to communicate to the service that the account was updated so we will get an updated view of the account's data.

**Comment by rkomandu** [comment]
@romayalon , are we planning to fix this in 5.15.3/.4

**Comment by anandhu-karattu** [comment]
I have a use case:
1. Update newbuckets path in the account. old_bucketspath to new_bucketspath
2. Create a new bucket via aws mb or s3api bucket-create, this is still creating new buckets under old_bucketspath
3. After 5-10 mins and account cache expiry, new buckets are getting created under new_bucketspath as expected.

So the issue I am facing here is , I need to wait 5-10 mins to test my scenarios or even in test automation.
Do we have a way to trigger account cache expiry manually? So that as a work around I can trigger it and proceed with my next steps?
@romayalon @guymguym

**Comment by romayalon** [comment]
@anandhu-karattu We do have process.env.ACCOUNTS_CACHE_EXPIRY which you can set on tests to be 1 ms, Can you try it? (add it to /etc/sysconfig/noobaa env file)
we can consider exporting it as a config as a W/A for customers who will complain the same issue, like we have config.OBJECT_SDK_BUCKET_CACHE_EXPIRY_MS...
Adding a point regarding IAM, IAM server will be on the same process as the service so we can easily invalidate accounts cache when updating accounts, this case can be a motivation for exploiting the tags for updating accounts config...
@guymguym @nimrod-becker @shirady 
WDYT?

**Comment by romayalon** [comment]
Updating that we had an internal discussion, we wouldn't recommend exporting it to customers as it could create very wasteful config files reads.
@anandhu-karattu as we discsussed on slack, you can use it on tests if you wish as we do.

**Comment by anandhu-karattu** [comment]
@romayalon I verified the method you suggested . Please find my observations:
1. Create a new file with config : echo ACCOUNTS_CACHE_EXPIRY=1 > /etc/sysconfig/noobaa
2. Restart the S3 service . 

Observations:
1. On a 2 node cluster, I updated this on node1. This parameter is effective after restart of services and right after account update (for --accesskeys as well as --newbucketspath) I could find the new value is effective.
2. After the test , I deleted the file “etc/sysconfig/noobaa” and restarted S3 service. Now the account expiry is back to normal 10 mins.

However this requires a restart of S3 service and we don't want to keep this config file option for all the tests. 
I suspect, I might miss some issues with this config.

**Instead do we have any way to call the server and refresh the cache individually ? If yes, We can even include that call in our account update cli (mms3) and refresh the cache after each update operation? Will that be possible?**

**Comment by romayalon** [comment]
Hey @anandhu-karattu 
If there was an easy option for invalidating the cache from outside of NooBaa endpoint process we would do it from our CLI...
As I wrote before, currently noobaa endpoint process supports only S3 operations (invalidate cache is a NooBaa internal function and not S3 request, so it's not supported) but in the future (currently WIP) the endpoint 
process will receive IAM requests for updating access keys (and invalidating the cache), and we need to consider supporting different configurations like new_buckets_path.. 
CC: @guymguym @shirady @nimrod-becker 

> @anandhu-karattu We do have process.env.ACCOUNTS_CACHE_EXPIRY which you can set on tests to be 1 ms, Can you try it? (add it to /etc/sysconfig/noobaa env file) we can consider exporting it as a config as a W/A for customers who will complain the same issue, like we have config.OBJECT_SDK_BUCKET_CACHE_EXPIRY_MS... Adding a point regarding IAM, IAM server will be on the same process as the service so we can easily invalidate accounts cache when updating accounts, this case can be a motivation for exploiting the tags for updating accounts config... @guymguym @nimrod-becker @shirady WDYT?

**Comment by romayalon** [comment]
@anandhu-karattu Could you set `echo ACCOUNTS_CACHE_EXPIRY=1 > /etc/sysconfig/noobaa` at the beginning of the tests (before enabling noobaa service) like we do instead of setting it for a specific test and restart the service?

**Comment by anandhu-karattu** [comment]
@romayalon I can set the parameter as we discussed and manage my tests (if there is no other way).
@madhuthorat how we can handle this use case ? Can we advise the customers to wait for 10 mins to create a new bucket (via s3 mb or s3api) after account update with --newbucketspath ?

**Comment by shirady** [comment]
Hi,
@anandhu-karattu we added a fix that would stat the config file of the account (it is merged to master).
Adding a label to validate it.

**Comment by rkomandu** [comment]
@shirady , master means here 4.18.x ?

**Comment by shirady** [comment]
@rkomandu Yes, in the upstream version of 4.18:
The fix was added in PR #8585.
You can see the commit hash b2ce22412b7220ae0b96111f1b2ef2a9bab35116 with the title "NC | NSFS | Add stat to account_cache"
And you can find this title in [5.18 commits](https://github.com/noobaa/noobaa-core/commits/5.18/)
I hope it helps.
